### PR TITLE
Storage manager

### PR DESCRIPTION
**Pull request by YomGarcia**
Minor fixes to StorageManagers:
- removed unused BinaryStorageManager
- added check on buffer size to allocate size-appropriate dataspace
- added xml attribute associated to the type of StorageManager used to generate the xml

### CODE REVIEW

**Review by jschueller — APPROVED** [review]

**Comment by YomGarcia** [comment]
closes #1801

File: python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py [hunk 1]
```diff
@@ -71,6 +71,7 @@
 #
 # - its name: it is useful to give names to the objects we want to save. If no name has been given by the user, we can use the name by default. The name of each object saved can be checked in the file.XML created or by printing the study in the python interface (with the command print).
 # - its id number : this id number is unique for each object. It is useful to distinguish objects of same type which names are identical, equal to the default name given ('Unnamed'). This id number may be checked by printing the study loaded in the python interface (with the command print): be careful, this print operation must be performed after having loaded the study (the id number may be different from the one indicated in the file.XML associated to the study).
+# - for HDF5 storage : the id is used to name the HDF5 dataset. This way, the various floating point/integer collections are stored and accessed through the name of their respective datasets.
```
Comment (josephmure) [inline python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py:74]:
```suggestion
# - for HDF5 storage: the id is used to name the HDF5 dataset. This way, the various floating point/integer collections are stored and accessed through the name of their respective datasets.
```
Does this mean that the id serves as both id and name, or that the HDF5 dataset has no name?

File: python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py [hunk 2]
```diff
@@ -81,6 +82,12 @@
 fileName = 'study.xml'
 study.setStorageManager(ot.XMLStorageManager(fileName))
 
+# %%
+# Alternatively, large amount of data can be stored in binary HDF5 file.; XML file ('study_h5.xml') serves as header for binary data (stored in 'study_h5.h5' which is automatically created).
```
Comment (josephmure) [inline python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py:86]:
```suggestion
# Alternatively, large amounts of data can be stored in binary HDF5 file. An XML file (`study_h5.xml`) serves as header for binary data, which are stored in the automatically created `study_h5.h5` file.
```

**Review by josephmure — COMMENTED** [review]
This PR is a good opportunity to improve the wording of the *Save/load a study* example. If you agree with my suggestions, you can commit them directly from the GitHub interface.

**Comment by YomGarcia** [comment]
> This PR is a good opportunity to improve the wording of the _Save/load a study_ example. If you agree with my suggestions, you can commit them directly from the GitHub interface.

Sure your changes are more than welcome

**Comment by YomGarcia** [comment]
I Included comments from Joseph and added an attribute to the xml file to check wether the appropriate StorageManager is being used when loading a Study. If the attribute is missing (backward compatibility issue), it is added on the fly when the xml is read

File: python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py (hunk 1 above)
Comment (josephmure) [reply python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py]: So does the HDF5 dataset have no name?

Comment (YomGarcia) [reply python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py]: The dataset gets named after the id, as shown in the code snippet below. id role in the xml remains unchanged
```  
<object class="Point" id="9" size="1" StudyVisible="false" StudyLabel="">
    <string>/Study_0.h5:/9</string>
  </object>```

Comment (YomGarcia) [reply python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py]: I updated this part to better reflect this

Comment (josephmure) [reply python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py]: Thanks for your explanation.

File: python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py [hunk 3]
```diff
@@ -17,7 +17,7 @@
 Save/load a study
 =================
 """
-# %% 
+# %%
```
Comment (josephmure) [inline python/doc/examples/numerical_methods/general_methods/plot_study_save_load.py:21]:
```suggestion
```
This empty cell can be removed.

**Review by josephmure — APPROVED** [review]

**Comment by jschueller** [comment]
@sofianehaddad you'll probably want to look into it even if we tested this already in persalys

**Review by regislebrun — APPROVED** [review]

**Review by regislebrun — APPROVED** [review]
The parenthesis of the if condition is so tightly packed to the keyword that it makes it looks like a function. A unique space would be appreciated to conform to the other parts of the code.

**Review by josephmure — CHANGES_REQUESTED** [review]
Upon further reflexion, I don't think this PR can close #1801 unless you also create an example in the API doc page:

https://github.com/openturns/openturns/blob/f1fce95b1485bd48a246a8f1d41351e4aa957891/python/src/Study_doc.i.in#L28
Here is the current output:

http://openturns.github.io/openturns/1.16/user_manual/_generated/openturns.Study.html

**Comment by YomGarcia** [comment]
> The parenthesis of the if condition is so tightly packed to the keyword that it makes it looks like a function. A unique space would be appreciated to conform to the other parts of the code.

It is a convention I tend to forget about, I have put some breathing room around those parenthesis

**Comment by YomGarcia** [comment]
> Upon further reflexion, I don't think this PR can close #1801 unless you also create an example in the API doc page:
> 
> https://github.com/openturns/openturns/blob/f1fce95b1485bd48a246a8f1d41351e4aa957891/python/src/Study_doc.i.in#L28
> 
> Here is the current output:
> 
> http://openturns.github.io/openturns/1.16/user_manual/_generated/openturns.Study.html

I wanted to avoid duplicate examples. Apart from setting a different instance of StorageManager, the rest is transparent to the user as far a reading and writing study files go. That is why I just added these lines (during the initial hdf5 PR):

https://github.com/openturns/openturns/blob/f1fce95b1485bd48a246a8f1d41351e4aa957891/python/src/Study_doc.i.in#L39

**Comment by josephmure** [comment]
Makes sense to me. What do you think @mbaudin47?

**Comment by josephmure** [comment]
@jschueller @YomGarcia I now run into an issue with `plot_study_save_load.py`. Curiously, this does not seem to happen to CI builds. Perhaps it depends on the installed dependencies?

```
WARNING: python/src/sphinx_build/examples/numerical_methods/general_methods/plot_study_save_load.py failed to execute correctly: Traceback (most recent call last):
  File "python/src/sphinx_build/examples/numerical_methods/general_methods/plot_study_save_load.py", line 87, in <module>
    study_h5.setStorageManager(ot.XMLH5StorageManager(fileName_h5))
TypeError: XMLH5StorageManager() takes no arguments
```

**Comment by jschueller** [comment]
yes, it fails without hdf5, I'm fixing it

**Comment by josephmure** [comment]
I still get the error, even after your fix.

**Comment by jschueller** [comment]
strange, it fixed the error for me

**Comment by josephmure** [comment]
Yes I did.

**Comment by jschueller** [comment]
do other hdf5 tests pass ?

**Comment by josephmure** [comment]
How can I see that?

**Comment by josephmure** [comment]
Here is the output of the CMake command:
```
devel@OT-VM:~/GIT/openturns/build-master-doc$ cmake .. -DCMAKE_INSTALL_PREFIX=$PWD/install -DPYTHON_EXECUTABLE=/usr/bin/python3 -DSPHINX_FLAGS="-j6"
-- A library with LAPACK API found.
-- Could NOT find Spectra (missing: SPECTRA_INCLUDE_DIRS) 
-- Found HMAT: /usr/local/lib/cmake/hmat
-- Could NOT find HDF5 (missing: HDF5_LIBRARIES HDF5_INCLUDE_DIRS CXX) (found version "")
-- Boost version: 1.67.0
-- Could NOT find MPFR (missing: MPFR_LIBRARY MPFR_INCLUDE_DIRS) 
-- Could NOT find MPC (missing: MPC_LIBRARY MPC_INCLUDE_DIRS) 
-- nlopt::GN_AGS: fails to build
-- Found dlib: /usr/lib/x86_64-linux-gnu/cmake/dlib (found version 19.10.0)
-- Found installed version of Eigen: /usr/lib/cmake/eigen3
-- Found required Ceres dependency: Eigen version 3.3.7 in /usr/include/eigen3
-- Found required Ceres dependency: glog
-- Found installed version of gflags: /usr/lib/x86_64-linux-gnu/cmake/gflags
-- Detected gflags version: 2.2.2
-- Found required Ceres dependency: gflags
-- Found Ceres version: 1.14.0 installed in: /usr with components: [EigenSparse, SparseLinearAlgebraLibrary, LAPACK, SuiteSparse, CXSparse, SchurSpecializations, OpenMP, Multithreading]
-- Could NOT find Bonmin (missing: COIN_BONMIN_LIBRARY COIN_INCLUDE_DIR) 
-- Could NOT find Ipopt (missing: IPOPT_LIBRARIES IPOPT_INCLUDE_DIRS) 
-- French locale found
-- std::regex found
-- Could NOT find chaospy (missing: CHAOSPY_LOCATION _CHAOSPY_VERSION_MATCH) 
-- Could NOT find dill (missing: DILL_LOCATION _DILL_VERSION_MATCH) 
-- The following OPTIONAL packages have been found:

 * BISON (required version >= 2.4)
 * FLEX
 * TBB
 * HMAT (required version >= 1.6)
 * LibXml2
 * Boost (required version >= 1.46)
 * NLopt
 * dlib
 * Ceres
 * CMinpack
 * OpenMP
 * SWIG (required version >= 3)
 * PythonInterp
 * PythonLibs (required version == 3.7.3)

-- The following REQUIRED packages have been found:

 * BLAS
 * LAPACK
 * Threads

-- The following OPTIONAL packages have not been found:

 * Spectra
 * HDF5
 * MPFR
 * MPC
 * Bonmin
 * Ipopt

-- Configuring done
-- Generating done
```

**Comment by josephmure** [comment]
After that, the compilation of the doc fails with the error I reported.

**Comment by jschueller** [comment]
ah, I fixed only the test, not the doc

**Pull request merged**